File: Mushroom/Grid.cs
using System.Collections.Generic;
using Godot;
using Mushroom.Ceils;
using Mushroom.Data;

namespace Mushroom;

public static class Grid
{
    private static CellBase[] _cells;
    public static Vector2I Size { get; private set; }
    public static int GroundLevel { get; internal set; }

    public static void Initialize(Vector2I size)
    {
        Size = size;
        _cells = new CellBase[size.X * size.Y];
        for (int i = 0; i < _cells.Length; i++)
            _cells[i] = Air.Instance;
    }

    public static void Set(Vector2I vector2, CellBase cell) => Set(vector2.X, vector2.Y, cell);
    
    public static void Set(int x, int y, CellBase cell)
    {
        if (IsInBounds(x, y))
            _cells[y * Size.X + x] = cell;
    }

    public static CellBase Get(Vector2I vector2) => Get(vector2.X, vector2.Y);
    
    public static CellBase Get(int x, int y)
    {
        if (IsInBounds(x, y))
            return _cells[y * Size.X + x] ?? Air.Instance;
        return Air.Instance;
    }

    public static void Move(Vector2I from, Vector2I to)
    {
        var target = Get(from.X, from.Y);
        Set(from, Get(to.X, to.Y));
        Set(to, target);
    }

    public static CellBase GetNeighbor(Vector2I vector2, Direction direction)
    {
        return direction switch
        {
            Direction.Up => Get(vector2.X, vector2.Y - 1),
            Direction.Down => Get(vector2.X, vector2.Y + 1),
            Direction.Left => Get(vector2.X - 1, vector2.Y),
            Direction.Right => Get(vector2.X + 1, vector2.Y),
            _ => Air.Instance
        };
    }

    public static bool IsInBounds(Vector2I vector2) => IsInBounds(vector2.X, vector2.Y);
    public static bool IsInBounds(int x, int y) => x >= 0 && x < Size.X && y >= 0 && y < Size.Y;
}

File: Core/src/grid.rs
pub mod grid {
    use std::collections::HashMap;
    use std::sync::Mutex;
    use lazy_static::lazy_static;
    use crate::cells::Cell;
    use crate::data::*;

    struct WorldState {
        cells: HashMap<(i32, i32), Box<dyn Cell>>,
        size: Size,
        ground_level: i32,
    }

    lazy_static! {
                static ref WORLD: Mutex<WorldState> = Mutex::new(WorldState {
                    cells: HashMap::new(),
                    size: Size { x: 0, y: 0 },
                    ground_level: 0,
                });
            }

    pub fn set(position: Position, cell: Box<dyn Cell>) {
        let mut world = WORLD.lock().unwrap();
        world.cells.insert((position.x, position.y), cell);
    }

    pub fn get(position: Position) -> Box<dyn Cell> {
        let world = WORLD.lock().unwrap();
        world.cells
            .get(&(position.x, position.y))
            .cloned()
            .unwrap_or_else(|| Box::new(crate::cells::air::Air))
    }

    pub fn move_cell(from: Position, to: Position) {
        if from == to { return; }

        let mut world = WORLD.lock().unwrap();

        if let Some(cell_from) = world.cells.remove(&(from.x, from.y)) {
            if let Some(cell_to) = world.cells.remove(&(to.x, to.y)) {
                world.cells.insert((from.x, from.y), cell_to);
            }
            world.cells.insert((to.x, to.y), cell_from);
        }
    }

    pub fn get_neighbor(position: Position, direction: Direction) -> Box<dyn Cell> {
        let target_pos = match direction {
            Direction::Up => Position::new(position.x, position.y - 1),
            Direction::Down => Position::new(position.x, position.y + 1),
            Direction::Left => Position::new(position.x - 1, position.y),
            Direction::Right => Position::new(position.x + 1, position.y),
        };
        get(target_pos)
    }

    pub fn is_in_bounds(position: Position) -> bool {
        let world = WORLD.lock().unwrap();
        position.x >= 0 && position.x < world.size.x && position.y >= 0 && position.y < world.size.y
    }

    pub fn get_size() -> Size {
        WORLD.lock().unwrap().size
    }

    pub fn set_size(new_size: Size) {
        WORLD.lock().unwrap().size = new_size;
    }

    pub fn get_ground_level() -> i32 {
        WORLD.lock().unwrap().ground_level
    }

    pub fn set_ground_level(level: i32) {
        WORLD.lock().unwrap().ground_level = level;
    }
}